Sz issue/149 fix commit message to merge instead of close

simonzhu24 opened the pull request:
Using Github_API Gem to Merge PR

codenamev commented on lib/git_reflow/commands/setup.rb:
```diff
@@ -25,5 +25,12 @@
     GitReflow::Config.add "constants.minimumApprovals", ask("Set the minimum number of approvals (leaving blank will require approval from all commenters): "), local: reflow_options[:project_only]
     GitReflow::Config.add "constants.approvalRegex", GitReflow::GitServer::PullRequest::DEFAULT_APPROVAL_REGEX, local: reflow_options[:project_only]
     
+    say("Enter the follow information associated with your remote repository: (The information is used to run 'git reflow deliver'.")
```
What does this have to do with the new merge?

simonzhu24 replied:
In order to use the github_api gem, we need to provide the payload, which needs to owner and repository name as well as the auth-token in order to make the HTTP request. I decided to put these in the git config file and prompt users to enter these.

codenamev replied:
We are already using the `github_api` gem.  See the [GitHelpers.remote_user](https://github.com/reenhanced/gitreflow/blob/master/lib/git_reflow/git_helpers.rb#L12) and [GitHelpers.remote_repo_name](https://github.com/reenhanced/gitreflow/blob/master/lib/git_reflow/git_helpers.rb#L17) for how we're using those to [search for existing pull requests](https://github.com/reenhanced/gitreflow/blob/master/lib/git_reflow.rb#L115).  [GitReflow.git_server](https://github.com/reenhanced/gitreflow/blob/master/lib/git_reflow.rb#L187) will give you an instance you can use to interface with the current git server's API.

simonzhu24 replied:
Yes? I understand this. But I need to provide parameters to the `GitReflow::GitServer::GitHub.connection.pull_requests.merge` method and this is what I am prompting for?

codenamev replied:
In your `pull_requests.merge` you can use `GitReflow.remote_user` and `GitReflow.remote_repo_name` for `owner`/`repo`.

simonzhu24 replied:
Thanks! I can't believe I didn't see this before. Makes my prompts for user and repo name unnecessary.

codenamev replied:
Please remove this.  We already setup the user's oauth-token in the `setup` command.

codenamev replied:
@simonzhu24 ping to remove.

simonzhu24 replied:
Absolutely.

codenamev commented on lib/git_reflow/git_server/bit_bucket/pull_request.rb:
```diff
@@ -78,6 +78,24 @@ def approvals
           approved
         end
 
+        def merge_feature_branch(feature_branch_name, options = {})
```
I think this should be the default implementation and live in the base pull request (`lib/git_reflow/git_server/pull_request.rb`).  Also, I think it makes more sense to name this `merge!` (see below for why I'd like to see the bang)

simonzhu24 replied:
Sure, but I wanted to keep the method name the same to be consistent with it's original naming scheme (the original name was `merge_feature_branch`).

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,38 @@ def approved?
+          )
+
+          append_to_squashed_commit_message(message) if message.length > 0
+          return res
```
Since the response is specific to GitHub, I would rather see the handling of the failure status codes here.  If it is anything but `response.ok` we should raise an error and catch the exception elsewhere (by default, if the error is not caught, this will cause reflow to immediately exit, which I think is ideal).

simonzhu24 replied:
Sure

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,45 @@ def approved?
           end
         end
 
+        def merge_feature_branch(feature_branch_name, options = {})
```
I would still like to see these named `merge!`, as we can infer what the feature branch name is (`current_branch`), and the bang should be present so the user knows that it is dangerous (will raise an error).  Also, since we're now in the context of the pull request, the only option we'd need to look for is a `destination_branch` override.  Every other option can pull on information from the instance of itself.  I think it would make sense to have a new method here as well that generates the commit message so that we can slim down the complexity of this method.

simonzhu24 replied:
Ok.

simonzhu24 replied:
@codenamev I am not exactly sure what the format is to get the: `commit_title, commit_message, sha, commit_number` from `pull_request.rb`. I tried looking in the code for examples, but couldn't find any. Do you just access these attributes from `self.source_object[:title]`? I was just wondering if you could provide an example of this? 
Thanks so much!

codenamev replied:
We are currently using [method_missing](https://github.com/simonzhu24/gitreflow/blob/sz-issue/149-Fix-Commit-Message-to-merge-instead-of-close/lib/git_reflow/git_server/pull_request.rb#L136) to delegate methods to the `source_object`.  So anything you can call on a a `github_api` PR object, you can call in this context.  Reference the existing `deliver` method for how we're passing those options.

simonzhu24 replied:
@codenamev correct me if I am wrong, but it seems we still need to pass the `sha` in from `options`. I am not sure we can access `sha` from `source_object`

codenamev replied:
`this.head.sha` :wink:

simonzhu24 replied:
Finished. Will send a diff soon!

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,45 @@ def approved?
+            "squash" => true
+          }
+          
+          res = GitReflow::GitServer::GitHub.connection.pull_requests.merge(
```
We should be clear with our variable names.  Let's change `res` to `merge_response`

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,45 @@ def approved?
+            data
+          )
+
+          if res.code == "405"
```
The `github_api` gem wraps responses in a [Github::ResponseWrapper](https://github.com/piotrmurach/github/blob/master/lib/github_api/response_wrapper.rb).  We should only have to check for `merge_response.success?` (meaning `200`).  If it's not a success, we should raise a custom error with the message.  Honeybadger has a great article on [doing this in a simple way](http://blog.honeybadger.io/ruby-custom-exceptions/).  Let's throw it in a file called `lib/git_reflow/git_server/pull_request/merge_error.rb` and name the class `GitReflow::GitServer::PullRequest::MergeError` and include it in this file.  Then we can just have an unsuccessful `else` that raises our new custom error.  Something along the lines of:

```ruby
if merge_response.success?
  append_to_squashed_commit_message(message) if message.length > 0
else
  if merge_response.has_key?(:message)
    raise MergeError.new(merge_response.message)
  else
    raise MergeError.new(merge_response.to_s)
  end
end
```

simonzhu24 replied:
Thanks! This makes it clear!

codenamev commented on lib/git_reflow/git_server/pull_request.rb:
```diff
@@ -140,6 +140,25 @@ def method_missing(method_sym, *arguments, &block)
         end
       end
+
+      def merge_feature_branch(feature_branch_name, options = {})
```
same here with the renaming to `merge!` with the adjusted params.

codenamev replied:
also, since you're removing the local syncs, you'll have to add them back to this method to ensure BitBucket continues to update the feature branch's dependencies.  These were 4 lines I found removed that we'll need to continue to support here, but just double check me :wink:

```ruby
update_current_branch
fetch_destination(base_branch)
update_destination(feature_branch)
update_destination(base_branch)
```

simonzhu24 replied:
Yup!

codenamev commented on lib/git_reflow.rb:
```diff
@@ -127,32 +123,36 @@ def deliver(options = {})
+              :message             => commit_message,
+              :head                => existing_pull_request.head
+            )
+          rescue StandardError => error
```
Change this to `rescue GitReflow::GitServer::PullRequest::MergeError`.  Rescuing from `StandardError` is typically a code-smell that points to fuzzy logic.

simonzhu24 replied:
Ok

simonzhu24 commented:
@codenamev Updated! Please let me know when you have a chance to check this out!

codenamev commented on README.rdoc:
```diff
@@ -205,45 +205,32 @@ You kick butt. You've got your code reviewed and now you're ready to merge it do
 We built this <b>to get in your way and make you follow the process</b>. If you don't like it, do it by hand. You already know what you're doing.
 
-You'll be presented with a prefilled commit message based on the body of your pull request which includes the text <code>Closes #XX</code> that will automatically close the associated pull request on github when deliver completes.
+You'll be presented with a prefilled commit message based on the body of your pull request which includes the text <code>Merges #XX</code> that will automatically merge the associated pull request on github when deliver completes.
```
This wording is now misleading.  Let's change to `You'll be presented with a prefilled commit message based on the body of your pull request with references to the pull request and reviewers.`

simonzhu24 replied:
Sure.

simonzhu24 replied:
Sure.

codenamev commented on lib/git_reflow.rb:
```diff
@@ -107,42 +108,34 @@ def deliver(options = {})
-          if committed
-            say "Merge complete!", :success
+          begin
+            res = existing_pull_request.merge!(
```
we don't need the assignment here

simonzhu24 replied:
Yup.

codenamev commented on lib/git_reflow.rb:
```diff
@@ -107,42 +108,34 @@ def deliver(options = {})
-            say "Merge complete!", :success
+          begin
+            res = existing_pull_request.merge!(
+              feature_branch,
```
we also don't need the `feature_branch` here as we can get this from within our `merge!` method with `GitReflow.current_branch`

simonzhu24 replied:
Done.

codenamev commented on lib/git_reflow.rb:
```diff
@@ -107,42 +108,34 @@ def deliver(options = {})
+          begin
+            res = existing_pull_request.merge!(
+              feature_branch,
+              :destination_branch  => base_branch
```
let's also use ruby 2 hash syntax wherever we can: `destination_branch: base_branch`

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow.rb:
```diff
@@ -107,42 +108,34 @@ def deliver(options = {})
+            )
+          rescue GitReflow::GitServer::MergeError => error
+              say error.inspect, :deliver_halted
+          else
```
having the `else` hanging around here after the `begin` block is a little strange.  Let's just throw everything in the `else` here into the `begin` section.

simonzhu24 replied:
Sure.

codenamev commented:
@simonzhu24 I am just realizing that by allowing Github to merge on our behalf, we are losing two features:

1. Aborting an over-eager or accidental delivery
2. Forced-delivery in cases where the CI server is acting up and the GH checks fail, but a delivery is still safe.

We will need to address these two features before we can merge.  For `2`, it may make sense to fallback to the previous behavior.  I haven't had time to think about how to address `1` outside of confirmation message.

Any ideas?

simonzhu24 commented:
During the manual merging process, `1` is something that we didn't really provide before. `ask "Would you like to push this branch to your remote repo and cleanup your feature branch? "` was the only check we had to ask if the users wanted to update their remote branch after the delivery and even then, we have already completed the delivery process and it was just whether they wanted to update their remote branch themselves. I could implement `1`, but in my opinion, this would be a new feature.

`2` is an interesting point. I don't believe our previous method of delivery checked CI results or specific github rejections either! It did, however, check if we were unable to create a commit. We did:

`committed = run_command_with_label 'git commit', with_system: true` and `if committed... else... end`. 

But I was not entirely sure which case would result in this case executing and if this is a case that we should be concerned about in our automatic merge process. 

@codenamev Just wondering if you had a different point of view?

codenamev commented:
> During the manual merging process, 1 is something that we didn't really provide before.

We consider delivery complete only once the remote base branch has been updated.  By asking for `Would you like to push this branch to your remote repo and cleanup your feature branch?`, we are able to pause the final push of the locally merged feature and let the user decide (as a final sanity check) that they really want to push to master.  We take the same level of caution pushing to master that we do deploying to production.  The `master` branch should be deployable at all times.  Some projects we even have triggers setup to automatically deploy to production when the remote master has been updated.  This feature is in place now, and I've had to use it on several occasions (and am very thankful it exists).

> 2 is an interesting point. I don't believe our previous method of delivery checked CI results or specific github rejections either! 

We currently have [this feature in place here](https://github.com/reenhanced/gitreflow/blob/master/lib/git_reflow.rb#L127).  The `good_to_merge?` method takes an option to force the merge.  If you have a CI server hooked up to Github, Github will not allow the merge via their API if the CI reports failures.  We have in the past worked with sensitive test suites where failures would only occur on the CI server and not locally.  In order to not hold up our progress, we put the option to force the merge in these special cases.  I'd have to get @nhance to approve removing the override here, but this would result in manual merges in these cases which could bring on more problems then we'd solve.

pboling commented:
@codenamev @nhance - I just went over this with @simonzhu24 and he had misunderstood the second issue you raised.  Now we are on the same page I think.  He is going to:
1. Add a confirmation step.
2. Add the old manual merge process as an optional fallback to the github_api process.

The confirmation step will also call the status code to display the state of the PR to the user so they can make an informed decision on the merge.

The fallback will be opt-in by the user.

nhance commented:
After setting up the restrictions for merging on github it would be cool if we can somehow enforce the rules you establish in github into reflow.

What I mean is github can refuse to allow a branch to merge when it fails whatever checks you've setup to be required. Instead of duplicating the logic into reflow, if we could somehow understand the settings are established inside of github, we can feed the error message back from github and make reflow a bit dumber, which is something both @codenamev and I always strive towards.

In other words, since administrators are allowed to force a branch to push, even if a check fails (assuming the settings allow it), it would be great to use github's concept of force-merging instead.

This makes reflow smarter about github's new restrictions, but dumber in regards to forcing a commit. Since we need to support bitbucket for a mostly consistent experience, it's hard to say which path is right.

simonzhu24 commented:
@pboling Thanks for the clarification. I will get this out today hopefully.

simonzhu24 commented:
@codenamev I'm ready!!!

![image](https://cloud.githubusercontent.com/assets/18274144/15033145/ad5b1dc8-121c-11e6-94d0-d6b0248acb04.png)

pboling commented on README.rdoc:
```diff
@@ -205,64 +205,52 @@ You kick butt. You've got your code reviewed and now you're ready to merge it do
-  Would you like to push this branch to your remote repo and cleanup your feature branch? y
-  Counting objects: 7, done.
-  Delta compression using up to 16 threads.
-  Compressing objects: 100% (4/4), done.
```
@simonzhu24 We should not make the default process fail.  If it fails because you are "currently on the branch" then perhaps we should switch to another branch to allow the delete?

Also the example in the readme should be happy path, showing everything working.

pboling commented on README.rdoc:
```diff
@@ -274,29 +262,43 @@ This is what we do behind the scenes when you do +deliver+
+
+    data = {
+      "commit_title",
+      "commit_message",
```
@simonzhu24 Do we still have the ability for the end user to provide a commit message that will be sent along via `github_api`?

simonzhu24 replied:
@pboling We don't prompt the user to specify a commit message for a specific `deliver` command. Instead the commit message is currently derived from the user's `review` command. It is `last commit message` by default, but you can specify any message you want with the `-m` flag.

We can make an extra prompt to make sure the user customizes the commit message. @pboling is this something that we want?

pboling replied:
Absolutely.  It is a big reason why we decided to use git reflow.  We want to have a custom deliver template that the user will be able to edit.  We want to retain this functionality in all scenarios.

simonzhu24 replied:
Sure. Will do.

pboling commented on README.rdoc:
```diff
@@ -274,29 +262,43 @@ This is what we do behind the scenes when you do +deliver+
+  If 'y', then we'll update the +base+ and delete the feature branch
+    
+    git pull origin #{base_branch}
+    git push origin #{base_branch}
```
This no longer happens when using the github API, right?  I suppose it would still happen in the fallback scenario for when github API fails to merge.

simonzhu24 replied:
@pboling In either the fallback scenario or bit bucket, after we squash-merged our feature branch into our base, I believe we do not push in `existing_pull_request.merge!`. So therefore, git push outside is necessary.

pboling replied:
@simonzhu24 Perhaps it would be good to note that this functionality "overview" is for the bitbucket standard, or github manual process, and the github manual process is only used as an optional fallback to the "remote" merge via the `github_api`.

simonzhu24 replied:
Ok, let me add that in quickly.

pboling commented on lib/git_reflow.rb:
```diff
@@ -107,53 +108,49 @@ def deliver(options = {})
+              if cleanup
+                # Pulls merged changes from remote base_branch
+                run_command_with_label "git pull origin #{base_branch}"
+                run_command_with_label "git push origin #{base_branch}"
```
I don't think we want to push again here, since the `existing_pull_request.merge!` should have already pushed in the fallback/bitbucket scenario, and the merge happens on the remote in the Github API scenario.

simonzhu24 replied:
@pboling same comment as before, if you look at the previous version of the code using the manual merging process, they had to do a git push after the merge, because they don't merge in the code?

pboling commented on lib/git_reflow/merge_error.rb:
```diff
@@ -0,0 +1,9 @@
+module GitReflow
+  module GitServer
+    class MergeError < StandardError
+      def initialize(msg="My default message")
```
Let's make the default error something more relevant, like `"Merge failed"`.

simonzhu24 replied:
Absolutely.

pboling commented:
@simonzhu24 Just three tiny comments!

codenamev commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;4kkC4aoH&#x2F;25-166-issue-149-fix-commit-message-to-merges-instead-of-closes"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> [166] [Issue #149] Fix Commit Message to &quot;Merges&quot; instead of &quot;Closes&quot;</a>

codenamev commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;cGptSZtu&#x2F;29-170-sz-issue-149-fix-commit-message-to-merge-instead-of-close"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> [170] Sz issue&#x2F;149 fix commit message to merge instead of close</a>

simonzhu24 commented:
@pboling @codenamev Let's try this again!

codenamev commented on lib/git_reflow.rb:
```diff
@@ -107,53 +108,58 @@ def deliver(options = {})
+          deliver = always_deliver || (ask "This is the current status of your Pull Request. Are you sure you want to deliver? ") =~ /^y/i
+          
+          if deliver
+            puts "Merging pull request ##{existing_pull_request.number}: '#{existing_pull_request.title}', from '#{existing_pull_request.feature_branch_name}' into '#{existing_pull_request.base_branch_name}'"
```
please use the `say` command instead of all these `puts`.  we have plans to replace `puts` with a logger and this will make it easier.

codenamev replied:
also, this should live inside the `merge!` method

simonzhu24 replied:
@codenamev can you be a bit more specific?

codenamev replied:
`say "Merging pull request ##{self.number}: '#{self.title}', from '#{self.feature_branch_name}' into '#{self.base_branch_name}'"` should be the first call inside the `merge!` methods

simonzhu24 replied:
Absolutely! Thanks!

codenamev commented on lib/git_reflow.rb:
```diff
@@ -107,53 +108,58 @@ def deliver(options = {})
+              if committed
+                say "Pull Request successfully merged.", :success
+
+                # check if user always wants to push and cleanup, otherwise ask
```
Let's throw this in a private method `cleanup_feature_branch?` so that we can _clean up_ the clutter in this method.  With it wrapped in a method, I'd also like to see the double assignment go away so we just have:

```ruby
  (
    Config.get('reflow.always-cleanup') == "true" ||
    (ask "Would you like to cleanup your feature branch? ") =~ /^y/i
  )
```

simonzhu24 replied:
This is beyond the scope of this diff. It's already getting a bit long, let's keep this diff for its original purpose.

codenamev replied:
it's easier to refactor in context and is only a few lines of code :wink:

simonzhu24 replied:
_I-can't-find-the-right-emoticon-right-now._

simonzhu24 replied:
Done.

codenamev commented on lib/git_reflow.rb:
```diff
@@ -107,53 +108,58 @@ def deliver(options = {})
+
+          # displays the current status and prompts user for confirmation
+          self.status base_branch
+          always_deliver = GitReflow::Config.get('reflow.always-deliver') == "true"
```
we are already in the context of `GitReflow` so we can just say `Config.get...` here

simonzhu24 replied:
Ok.

codenamev replied:
fair enough, let's just keep this syntax

codenamev commented on lib/git_reflow.rb:
```diff
@@ -107,53 +108,58 @@ def deliver(options = {})
         say "No pull request exists for #{remote_user}:#{current_branch}\nPlease submit your branch for review first with \`git reflow review\`", :deliver_halted
       else
-
         commit_message = if "#{existing_pull_request.description}".length > 0
```
Let's move the `commit_message` into the `merge!` command as well.

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow.rb:
```diff
@@ -107,53 +108,58 @@ def deliver(options = {})
+            puts "Merging pull request ##{existing_pull_request.number}: '#{existing_pull_request.title}', from '#{existing_pull_request.feature_branch_name}' into '#{existing_pull_request.base_branch_name}'"
+
+            begin
+              committed = existing_pull_request.merge!(destination_branch: base_branch, force: options['skip_lgtm'], message: commit_message)
```
With the `commit_message` living inside the `merge!` method, we don't need it provided as an option now.  It can be easily inferred in the context of the PR.

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow.rb:
```diff
@@ -107,53 +108,58 @@ def deliver(options = {})
+                say "There were problems commiting your feature... please check the errors above and try again.", :error
+              end
+            rescue GitReflow::GitServer::MergeError => error
+                say error.inspect, :deliver_halted
```
Let's make this `error.to_s`.  [StandardError#to_s](http://ruby-doc.org/core-2.3.1/Exception.html#method-i-to_s) will report the exception's `message` by default.

simonzhu24 replied:
Absolutely. This is a good point.

codenamev commented on lib/git_reflow/commands/review.rb:
```diff
@@ -3,8 +3,8 @@
   c.arg_name '[destination_branch] - the branch you want to merge your feature branch into'
-  c.flag [:t, :title], default_value: 'last commit message'
-  c.flag [:m, :message], default_value: 'title'
+  c.flag [:t, :title]
```
please pull in the latest `reenhanced/master` as we have fixed this

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow/git_server/pull_request.rb:
```diff
@@ -140,6 +140,33 @@ def method_missing(method_sym, *arguments, &block)
+      def merge!(options = {})
+
+        destination_branch = options[:destination_branch] || 'master'
+        message = "#{options[:message]}"
```
might as well add a method for this:

```ruby
def commit_message_for_merge
  if "#{self.description}".length > 0
    "#{self.description}"
  else
    "#{GitReflow.get_first_commit_message}"
  end
end

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,64 @@ def approved?
+
+          # fallback to default merge process if user "forces" merge
+          if(options[:force])
+            super options
```
:heart: this :smile:

codenamev commented on lib/git_reflow/git_server/pull_request.rb:
```diff
@@ -140,6 +140,33 @@ def method_missing(method_sym, *arguments, &block)
+        GitReflow.update_destination(feature_branch_name)
+        GitReflow.update_destination(destination_branch)
+
+        if "#{options[:pull_request_number]}".length > 0
```
we can then remove these message builder segments

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,64 @@ def approved?
+            options[:destination_branch] ||= 'master'
+            feature_branch_name = GitReflow.current_branch
+
+            commit_title = ask("Please enter your delivery commit title: (leaving blank will use default)")
```
I'm not really a fan of this double prompt.  I think we'd like to see this handled in [a similar fashion to the review command](https://github.com/reenhanced/gitreflow/blob/master/lib/git_reflow.rb#L53).  Ideally it would use the new `commit_message_for_merge` method provided below to populate the `"#{git_root_dir}/.git/SQUASH_MSG"` file and open it for the user to enter/edit their commit message using the `ENV['EDITOR']`.  Then once the use quits/saves we would just read that file and use the first line as the title and the rest of it for the body.

pboling replied:
I agree.

simonzhu24 replied:
Sure.

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,64 @@ def approved?
+              message << "\nLGTM given by: @#{lgtm_authors.join(', @')}\n"
+            end
+
+            data = {
```
I don't think we need this extra assignment as we only use it in one place below.

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,64 @@ def approved?
+              GitReflow.append_to_squashed_commit_message(message) if message.length > 0
+              return true
+            else
+              if merge_response.has_key?(:message)
```
turns out `StandardError#to_s` calls `message` by default, so we can get rid of the conditional and just `raise MergeError.new(merge_response.to_s)`

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,64 @@ def approved?
+
+            if merge_response.success?
+              GitReflow.run_command_with_label "git checkout #{options[:destination_branch]}"
+              GitReflow.append_to_squashed_commit_message(message) if message.length > 0
```
we don't need this any more :wink:

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,64 @@ def approved?
+            )
+
+            if merge_response.success?
+              GitReflow.run_command_with_label "git checkout #{options[:destination_branch]}"
```
let's also `GitReflow.run_command_with_label "git pull #{options[:destination_branch]}"` to grab the merge that was made.

simonzhu24 replied:
@codenamev I am running:

```
run_command_with_label "git pull origin #{base_branch}"
run_command_with_label "git push origin #{base_branch}"
```

in `GitReflow.rb` after the merge! if it is successful. So this is already covered!

simonzhu24 commented:
@codenamev This last PR incorporated quite a lot of changes:

1. Separation of logic into `cleanup_feature_branch?`, `deliver?`, `commit_message_for_merge` methods
2. Wrote tests for `cleanup_feature_branch?`, `deliver?`, `commit_message_for_merge` methods
3. Added tests for `merge!` exception (checked to make sure an exception was thrown)
4. `options` and `inputs` was getting confusing since some using symbol parameters and some are using string parameters, so I migrated everything in `review` and `deliver` commands to use symbols as a default. 
5. Added default editor way for entering `commit_title` and `commit_message` for `merge!` 
(In order to pass the rspec tests and make this segment testable, I need to use `options[:title]` and `options[:message]` to avoid bringing up the editor. So I needed to pass `options` to `merge`.)
6. Used `error.to_s` instead of `error.inspect`
7. Removed `commit_message` as an options input in `deliver` call to `merge!`
8. Moved "Merging Pull Request ..." to `merge!` in `pull_request.rb`

** Since we only need to pull and shouldn't push in the case that we use the github_api gem, so I moved pull && push to the `merge!` method of the super class of `pull_request` and kept the pull inside of `deliver` in `git_reflow.rb` to make sure our local `base_branch` is always up-to-date.

The `push` command shouldn't matter after merging because there normally wouldn't be a merge commit, but that is not always the case (the user can always decide to make their own commit during the process), so we should be careful.

codenamev commented on lib/git_reflow.rb:
```diff
@@ -118,42 +123,34 @@ def deliver(options = {})
+          # displays current status and prompts user for confirmation
+          self.status base_branch
+          
+          if self.deliver?
```
There is far too much nested logic going on here, and we've been looking for a good time to refactor this.  I think it would be better to move this `deliver?` check into the `merge!` method.

simonzhu24 replied:
Ok. I will rewrite this and update the tests to reflect this.

codenamev commented on lib/git_reflow.rb:
```diff
@@ -118,42 +123,34 @@ def deliver(options = {})
+          
+          if self.deliver?
+            begin
+              committed = existing_pull_request.merge!(options)
```
there's no need for the assignment if we're not using the `committed` variable elsewhere

simonzhu24 replied:
yup

codenamev commented on lib/git_reflow.rb:
```diff
@@ -118,42 +123,34 @@ def deliver(options = {})
+                if self.cleanup_feature_branch?
+                  run_command_with_label "git push origin :#{feature_branch}"
+                  run_command_with_label "git branch -D #{feature_branch}"
+                  puts "Nice job buddy."
```
`puts` => `say`

simonzhu24 replied:
@codenamev I was being consistent with the original version of the code. The code refactored used `puts` over `say` for this message with no color. So "Nice job buddy" used puts. Are you sure you want me to modify this?

codenamev replied:
We have a common practice of "refactoring in context".  Basically, since we're already modifying this code, it would make sense to refactor as long as the refactor stays within the context of the current scope.

simonzhu24 replied:
Sure!

codenamev commented on lib/git_reflow.rb:
```diff
@@ -118,42 +123,34 @@ def deliver(options = {})
+                  run_command_with_label "git branch -D #{feature_branch}"
+                  puts "Nice job buddy."
+                else
+                  puts "Cleanup halted.  Local changes were not pushed to remote repo.".colorize(:red)
```
Same here.  Also, in the context of using the new Github merge, this no longer applies.  It may be better/easier to make two new methods on PullRequest: 1) `cleanup_feature_branch`, 2) `cleanup_failure_message`.  Then from L137 to L144 would look like this:

```ruby
if existing_pull_request.cleanup_feature_branch
  say "Nice job buddy."
else
  say existing_pull_request.cleanup_failure_message, :deliver_halted
end
```

pboling replied:
👍 makes sense.

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,65 @@ def approved?
+            say "Merging pull request ##{self.number}: '#{self.title}', from '#{self.feature_branch_name}' into '#{self.base_branch_name}'"
+
+            options[:base] ||= 'master'
+            feature_branch_name = GitReflow.current_branch
```
We should be using `self.feature_branch_name` instead.

simonzhu24 replied:
Ok. This logic for `options[:base]` will be gone when we remove the `base` option altogether.

simonzhu24 replied:
Sure.

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,65 @@ def approved?
+
+              # prompts user for commit_title and commit_message
+              pull_request_msg_file = "#{GitReflow.git_root_dir}/.git/SQUASH_MSG"
+              default_editor        = "#{ENV['EDITOR']}"
```
This should move to a constant (now that we have a need to reference it in more than one place).  Put this in the main `lib/git_reflow.rb` file:

```ruby
DEFAULT_EDITOR = "#{ENV['EDITOR']}".freeze || "vi".freeze
```

pboling replied:
👍

simonzhu24 replied:
Ok.

simonzhu24 replied:
Sure.

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,65 @@ def approved?
+              pull_request_msg_file = "#{GitReflow.git_root_dir}/.git/SQUASH_MSG"
+              default_editor        = "#{ENV['EDITOR']}"
+
+              if default_editor.empty?
```
with the constant in use we won't need this check

simonzhu24 replied:
Sure.

simonzhu24 replied:
Ok.

codenamev commented on lib/git_reflow/git_server/git_hub/pull_request.rb:
```diff
@@ -78,6 +78,65 @@ def approved?
+              end
+
+              File.open(pull_request_msg_file, 'w') do |file|
+                file.write("#{self.title}\n#{self.body}\n")
```
if the `body` is blank we want to retain the previous behavior of using the first commit message

pboling replied:
👍

simonzhu24 replied:
Sure.

simonzhu24 replied:
Ok.

simonzhu24 commented:
@codenamev A question for you. From the context of the PR, we have the information for the `base_branch` that we want to `merge` into. However, upon `delivering,` we are asking the user to pass us an optional `base_branch` for the merge and if none is provided, we just merge into `master` (which could be a bit of an issue if the `base_branch` of the PR is not actually `master.` 

Also, hypothetically, it is possible for a user to get an LGTM for a PR merging into branch_A and then the user delivering into branch_B. 

Are either of these issues that we should be concerned about? 

If so, we can remove the option for users to provide a `base_branch` on deliver and just use the base_branch that we have from the PR. We needed this option before because all of the previous logic was in `git_reflow.rb`, from which context we didn't know which `base_branch` we were going to merge into. Now that we are in `pull_request.rb`, we have this information and no longer need to pass in the `base.`

codenamev commented:
@simonzhu24 Yeah, from the perspective of an existing pull request, we shouldn't be allowed to provide a different base branch.  Let's remove that from the options of the `merge!` method.  We're already validating the user's choice of a custom base branch with `find_open_pull_request` at the very beginning of the deliver command.

simonzhu24 commented:
@codenamev The question is why are we prompting the user to enter a `base_branch` in the first place? Can't we just use the `base_branch` for the PR?

codenamev commented:
Ah I see.  Looks like that has been around since the beginning, but I can't think of a scenario we've ever used it.  Let's pull it out of there! :smile:

codenamev commented:
Just want to say how excited we are to see this and to get it merged.  Thanks for all the hard work @simonzhu24!  Lots of great improvements in this.

simonzhu24 commented:
@codenamev Can you please review this new version? I would like to establish a path to getting this PR closed, as I have due dates I need to make and the continuous refactoring, while useful from a code quality standpoint, is impacting my ability to do so. I'm strongly in support of code readability and quality, but would prefer if we started handling these refactor tasks via separate issues so that we can ship the functionality, which we have developers waiting to use.

codenamev commented:
LGTM 😂 

Thanks again for this @simonzhu24!  It is hard for us to budge from our internal processes, but we will try to be gentler in your future contributions :wink:.  The changes you've made here surround a sensitive and important part of the code base which is why we have been so particular.
